File: packages/javascript/lib/date/createDateListFromRange.ts
import { dateToYYYYMMDD } from './dateToYYYYMMDD.js';
import { isValidDate } from './isValidDate.js';

export function createDateListFromRange(startDate: Date, endDate: Date): string[] {
	if (!isValidDate(dateToYYYYMMDD(startDate)) || !isValidDate(dateToYYYYMMDD(endDate))) {
		throw new Error('Either the start date or end date is an invalid date.');
	}

	const dateDiff = Math.floor(
		(Date.parse(endDate.toString()) - Date.parse(startDate.toString())) / 86400000
	);

	if (dateDiff < 0) {
		throw new Error('The start date is more recent than the end date.');
	}

	const endDateStr: string = dateToYYYYMMDD(endDate);
	const allDate: string[] = [];
	let currentYear: number = startDate.getFullYear();
	let currentMonth: number = startDate.getMonth() + 1;
	let currentDay: number = startDate.getDate();
	let currentDateStr = '';

	const createNewDateStr = (year: number, month: number, day: number): string =>
		`${year}-${month < 10 ? '0' : ''}${month}-${day < 10 ? '0' : ''}${day}`;

	while (endDateStr !== currentDateStr) {
		if (/[0-9]{4}-12-31/g.test(currentDateStr)) {
			currentYear += 1;
			currentMonth = 1;
			currentDay = 1;
		}

		const currentNewDateStr = createNewDateStr(currentYear, currentMonth, currentDay);

		if (isValidDate(currentNewDateStr)) {
			currentDay += 1;
			allDate.push(currentNewDateStr);
			currentDateStr = currentNewDateStr;
		} else {
			currentMonth += 1;
			currentDay = 1;
			currentDateStr = createNewDateStr(currentYear, currentMonth, currentDay);
		}
	}

	return allDate;
}

File: packages/python/qsu/date/createDateListFromRange.py
import math
import re
from datetime import datetime

from .dateToYYYYMMDD import dateToYYYYMMDD
from .isValidDate import isValidDate

_YEAR_END_RE = re.compile(r'[0-9]{4}-12-31')


def createDateListFromRange(startDate: datetime, endDate: datetime):
	if not isValidDate(dateToYYYYMMDD(startDate)) or not isValidDate(
		dateToYYYYMMDD(endDate)
	):
		raise ValueError('Either the start date or end date is an invalid date.')

	date_diff = math.floor((endDate.timestamp() - startDate.timestamp()) / 86400)

	if date_diff < 0:
		raise ValueError('The start date is more recent than the end date.')

	end_date_str = dateToYYYYMMDD(endDate)
	all_date = []
	current_year = startDate.year
	current_month = startDate.month
	current_day = startDate.day
	current_date_str = ''

	def create_new_date_str(year: int, month: int, day: int) -> str:
		m = f'0{month}' if month < 10 else f'{month}'
		d = f'0{day}' if day < 10 else f'{day}'
		return f'{year}-{m}-{d}'

	while end_date_str != current_date_str:
		if _YEAR_END_RE.search(current_date_str):
			current_year += 1
			current_month = 1
			current_day = 1

		current_new_date_str = create_new_date_str(
			current_year, current_month, current_day
		)

		if isValidDate(current_new_date_str):
			current_day += 1
			all_date.append(current_new_date_str)
			current_date_str = current_new_date_str
		else:
			current_month += 1
			current_day = 1
			current_date_str = create_new_date_str(
				current_year, current_month, current_day
			)

	return all_date
